packages/shared-metrics/src/gc.js

```diff
@@ -37,37 +48,46 @@ const nativeModuleLoader = require('./util/nativeModuleRetry')({
 
 let senseIntervalHandle;
 
-exports.activate = function activate() {
-  nativeModuleLoader.on('loaded', gcStats => {
-    exports.currentPayload.statsSupported = true;
-    gcStats.on('stats', stats => {
-      // gcstats exposes start and end in nanoseconds
-      const pause = (stats.end - stats.start) / 1000000;
-      gcPauseWindow.addPoint(pause);
-      const type = stats.gctype;
-      if (type === 1) {
-        minorGcWindow.addPoint(1);
-      } else if (type === 2) {
-        majorGcWindow.addPoint(1);
-      } else if (type === 4) {
-        incrementalMarkingsWindow.addPoint(1);
-      } else if (type === 8) {
-        processWeakCallbacksWindow.addPoint(1);
-      } else if (type === 15) {
-        minorGcWindow.addPoint(1);
-        majorGcWindow.addPoint(1);
-        incrementalMarkingsWindow.addPoint(1);
-        processWeakCallbacksWindow.addPoint(1);
-      }
-      exports.currentPayload.usedHeapSizeAfterGc = stats.after.usedHeapSize;
-    });
-    startSensing();
-  });
+nativeModuleLoader.on('loaded', gcStats_ => {
+  gcStats = gcStats_;
+  exports.currentPayload.statsSupported = true;
+  if (activateHasBeenCalled) {
+    actuallyActivate();
+  }
+});
+
+nativeModuleLoader.on('failed', () => {
+  exports.currentPayload.statsSupported = false;
+});
 
-  nativeModuleLoader.on('failed', () => {
-    exports.currentPayload.statsSupported = false;
+function actuallyActivate() {
+  if (hasBeenActivated) {
+    return;
+  }
+  hasBeenActivated = true;
+  gcStats.on('stats', stats => {
+    // gcstats exposes start and end in nanoseconds
+    const pause = (stats.end - stats.start) / 1000000;
+    gcPauseWindow.addPoint(pause);
+    const type = stats.gctype;
+    if (type === 1) {
+      minorGcWindow.addPoint(1);
+    } else if (type === 2) {
+      majorGcWindow.addPoint(1);
+    } else if (type === 4) {
+      incrementalMarkingsWindow.addPoint(1);
+    } else if (type === 8) {
+      processWeakCallbacksWindow.addPoint(1);
+    } else if (type === 15) {
+      minorGcWindow.addPoint(1);
+      majorGcWindow.addPoint(1);
+      incrementalMarkingsWindow.addPoint(1);
+      processWeakCallbacksWindow.addPoint(1);
+    }
+    exports.currentPayload.usedHeapSizeAfterGc = stats.after.usedHeapSize;
```

Comment:
~~Is `stats.after.usedHeapSize` a function or an attribute that will be read only once?~~
~~As this is inside of the `on` event, if this is an attribute, I am honestly not entirely sure if Node will update an exported value or what could be the implications with that. Could this be an issue?~~

I saw that something similar is done already at line 94 to 97, so I think this just works as intended.

Reply:
We read the exports.currentPayload once per second here: https://github.com/instana/nodejs-sensor/blob/7abc5a5fd3c74bd3a9592decf3efa700f3d3a292/packages/core/src/metrics/index.js#L58 before processing it further.